### ISSUE TITLE
frequent OSError: bad file descriptor 

### ISSUE THREAD

**Issue by fmv1992** [xonsh/xonsh#2496]
Hello.

I've been getting a lot of bad file descriptors lately:

~~~~~~
Exception in thread Thread-53:
Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/site-packages/xonsh/proc.py", line 1331, in run
    sp_stdout = io.TextIOWrapper(io.open(self.c2pwrite, 'wb', -1),
OSError: [Errno 9] Bad file descriptor
~~~~~~

It happens erratically. I'll be posting in other contexts and and with different error messages as they appear.

I have tried a different version of xonsh and prompt_toolkit but none worked.

xonfig:
~~~~~~
+------------------+---------------------------------------------------+
| xonsh            | 0.5.9                                             |
| Git SHA          | 2ae5218                                           |
| Commit Date      | Mar 15 12:06:26 2017                              |
| Python           | 3.6.2                                             |
| PLY              | 3.9                                               |
| have readline    | True                                              |
| prompt toolkit   | 1.0.7                                             |
| shell type       | prompt_toolkit                                    |
| pygments         | 2.2.0                                             |
| on posix         | True                                              |
| on linux         | <xonsh.lazyasd.LazyBool object at 0x7f49165ff860> |
| distro           | arch                                              |
| on darwin        | False                                             |
| on windows       | False                                             |
| on cygwin        | False                                             |
| is superuser     | False                                             |
| default encoding | utf-8                                             |
| xonsh encoding   | utf-8                                             |
| encoding errors  | surrogateescape                                   |
+------------------+---------------------------------------------------+
~~~~~~

**Comment by gforsyth** [comment]
Hi @fmv1992 -- thanks for reporting -- could you please update to the latest version and see if the issue keeps persisting?  Thanks

**Comment by fmv1992** [comment]
Do you mean the `xonsh-0.5.12` version or the development version?

In the case of the former I have already tried. Actually I have the impression that this error was more frequent than it is on `xonsh-0.5.9`.

In the case of the latter, feel free to indicate a specific commit to try out

: )

**Comment by gforsyth** [comment]
HI @fmv1992 -- I meant either, generally it's just good to know if it's still an issue on the latest release or the dev version and the easiest way to check that is just to update.

Strange that, as you put it, they started popping up "lately" -- was this not an issue before?
Could you try it out in an environment with python 3.5 or maybe 3.6.1 and see if the issue persists?  I imagine that it will, but if that's the only thing that's changed...

Thanks!

**Comment by melMass** [comment]
@fmv1992 Does this happen with `xonsh --no-rc` ? 
I had the same issue with some xontribs.

**Comment by astronouth7303** [comment]
@melMass which ones?

**Comment by melMass** [comment]
@astronouth7303 check #2510 !

**Comment by scopatz** [comment]
So is this a powerline issue?

**Comment by fmv1992** [comment]
@melmass,

I have not tested I'm sorry. Thus I am not able to say precisely what it is but I know its there. It happens on almost every sessions, specially in initialization.

**Comment by fmv1992** [comment]
This is still plaguing me every day. Just a refresher:

~~~~~~
Traceback (most recent call last):
  File "<frozen importlib._bootstrap_external>", line 833, 
OSError: [Errno 9] Bad file descriptor

During handling of the above exception, another exception o

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/xonsh/main.py", li
    return main_xonsh(args)
  File "/usr/lib/python3.6/site-packages/xonsh/main.py", li
    shell.shell.cmdloop()
  File "/usr/lib/python3.6/site-packages/xonsh/ptk/shell.py
    line = self.singleline(auto_suggest=auto_suggest)
  File "/usr/lib/python3.6/site-packages/xonsh/ptk/shell.py
    self.styler.style_name = env.get('XONSH_COLOR_STYLE')
  File "/usr/lib/python3.6/site-packages/xonsh/base_shell.p
    from xonsh.pyghooks import XonshStyle
  File "/usr/lib/python3.6/site-packages/xonsh/pyghooks.py"
    from pygments.lexers.agile import PythonLexer
  File "/usr/lib/python3.6/site-packages/pygments/lexers/ag
    from pygments.lexers.perl import PerlLexer, Perl6Lexer
  File "<frozen importlib._bootstrap>", line 961, in _find_
  File "<frozen importlib._bootstrap>", line 950, in _find_
  File "<frozen importlib._bootstrap>", line 655, in _load_
  File "<frozen importlib._bootstrap_external>", line 674, 
  File "<frozen importlib._bootstrap_external>", line 780, 
  File "<frozen importlib._bootstrap_external>", line 833, 
OSError: [Errno 9] Bad file descriptor
Xonsh encountered an issue during launch
Failback to /bin/sh
~~~~~~

Other shell's languages are just too insane for me to consider them. But please fix this : (

**Comment by scopatz** [comment]
Hi @fmv1992 - Sorry that this is still an issue :(.  Thanks for posting the traceback today.  That seems very different than the one in the initial post.

**Comment by fmv1992** [comment]
@scopatz,

I'm glad to be of any help. And I'm thankful for xonsh ever existing.

In fact the error messages are very diverse. There it goes another one; I sucessfully opened 3 shells before hitting it again:

~~~~~~
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/xonsh/ptk/shortcuts.py", line 109, in prompt
    document = cli.run(reset_current_buffer=False)
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/interface.py", line 389, in run
    self.eventloop.run(self.input, self.create_eventloop_callbacks())
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/eventloop/posix.py", line 110, in run
    fds = self._ready_for_reading(remaining_timeout)
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/eventloop/posix.py", line 183, in _ready_for_reading
    fds = select_fds(read_fds, timeout)
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/eventloop/select.py", line 35, in select_fds
    result = _python3_selectors(read_fds, timeout)
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/eventloop/select.py", line 69, in _python3_selectors
    sel.register(fd, selectors.EVENT_READ, None)
  File "/usr/lib/python3.6/selectors.py", line 412, in register
    self._epoll.register(key.fd, epoll_events)
OSError: [Errno 22] Invalid argument
~~~~~~

**Comment by scopatz** [comment]
> I'm glad to be of any help. And I'm thankful for xonsh ever existing.

Thanks! :smile: 

Could you run `import resource; resource.RLIMIT_NOFILE` and let us know what you are seeing?

```
scopatz@artemis ~/xonsh moveback $ import resource 
scopatz@artemis ~/xonsh moveback $ resource.RLIMIT_NOFILE
7
```

**Comment by scopatz** [comment]
All of these errors do seem to be related to select() or poll()

**Comment by fmv1992** [comment]
Yes, there it goes:

~~~~~~
monteiro|home/monteiro|$ import resource ; print(resource.RLIMIT_NOFILE)
7
~~~~~~

I'm not sure if my `~/.xonshrc` may be triggering it.

**Comment by scopatz** [comment]
Can you post or send me your xonshrc, please?

**Comment by fmv1992** [comment]
oh my, it is big and it also sources other files

one moment

**Comment by fmv1992** [comment]
I will not post all the files here because it may contain sensitive material.
However I have sent in the email you provided on your github page.
Let me know your thoughts either here or via email (if the mistake is very specific to those files).

**Comment by scopatz** [comment]
Thanks for sending them along @fmv1992.  I am still not able to reproduce this exactly, but I am curious, does this happen if you use the readline shell, instead of prompt-toolkit?  Does this happen if you start xonsh with `--no-rc`, as suggested above?  Just trying to eliminate possibilities.

**Comment by fmv1992** [comment]
@scopatz,

Requested results:

1. It also raises errors with `readline` although less frequently (I would say
   1/10 versus 1/6 with `prompt_toolkit`).
1. If I launch xonsh via `/usr/bin/xonsh`, that is, within the same terminal
   session this does not happen for neither shell types. I have launched
   a series of /usr/bin/xonsh during a working session of xonsh. My terminal is
   uxterm.
1. I have opened around 40 different uxterm terminals without my rcfile (moved
   it to a different folder and the aesthetics of xonsh changed accordingly)
   and none of them gave me errors.

Thus we can almost certainly conclude that this is:

1. Indifferent to either `$SHELL_TYPE = 'prompt_toolkit'` and `$SHELL_TYPE
   = 'readline'`.
1. Related to opening a new terminal.
1. Possibly not related to my xonshrc as opening xonshs inside of xonshs (item
   2 above) does not trigger the issue. However the absence of my xonshrc
   indicates otherwise.

Hopefully the experiments and theirs results are clear.

Let me know if you need more info.

**Comment by scopatz** [comment]
I still couldn't isolate this.

**Comment by fmv1992** [comment]
No problem. I'm thankful for your efforts.

If I gather new relevant information I'll let you know.

**Comment by scopatz** [comment]
In other news #2559 may help with some of these issues...

**Comment by andreoliwa** [comment]
First of all, thanks for xonsh. :)

I'm also having some similar "Bad file descriptor" exceptions like this:

<details>

```
Exception in thread Thread-75:
Traceback (most recent call last):
  File "/usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/usr/local/Cellar/xonsh/0.8.7/libexec/lib/python3.7/site-packages/xonsh/__amalgam__.py", line 13364, in run
    io.open(self.c2pwrite, "wb", -1), encoding=enc, errors=err
OSError: [Errno 9] Bad file descriptor
```

</details>

I managed to almost eliminate the errors by using `source-bash --interactive False` when sourcing any bash script inside `.xonshrc`, as described in this other issue: https://github.com/xonsh/xonsh/issues/2790#issuecomment-416283371

I could not isolate which sourced script caused the problem; my files are somewhat big.

- [Here is my `.xonshrc`](https://github.com/andreoliwa/dotfiles/blob/master/roles/xonsh/files/xonshrc.xsh.link);
- [Here is my `.bashrc`](https://github.com/andreoliwa/dotfiles/blob/master/roles/bash/files/bashrc.link);
- Below are the outputs of `xonfig` and `xpip freeze`.
 
<details>

```
$ xonfig
+------------------+----------------------+
| xonsh            | 0.8.12               |
| Git SHA          | 986bf28d             |
| Commit Date      | Mar 11 14:37:52 2019 |
| Python           | 3.7.2                |
| PLY              | 3.11                 |
| have readline    | True                 |
| prompt toolkit   | 2.0.7                |
| shell type       | prompt_toolkit2      |
| pygments         | 2.3.1                |
| on posix         | True                 |
| on linux         | False                |
| on darwin        | True                 |
| on windows       | False                |
| on cygwin        | False                |
| on msys2         | False                |
| is superuser     | False                |
| default encoding | utf-8                |
| xonsh encoding   | utf-8                |
| encoding errors  | surrogateescape      |
+------------------+----------------------+
$ xpip freeze
prompt-toolkit==2.0.7
Pygments==2.3.1
PyYAML==3.13
setproctitle==1.1.10
six==1.12.0
wcwidth==0.1.7
xonsh==0.8.12
xonsh-autoxsh==0.3
xontrib-z==0.4
```

</details>


I hope this helps with more clues to solve the issue.

**Comment by scopatz** [comment]
oof sorry you were encountering this issue too @andreoliwa. I wish we could isolate it.

**Comment by bloff** [comment]
I'm also having this issue often. It is very annoying and is preventing me from adopting xonsh as my main shell.

Any easy way of setting up my computer so that I will automatically enter a debug shell when this error happens?

**Comment by gforsyth** [comment]
@bloff -- a few things to check (that may help flush this out?)

1. If you start `xonsh` from `bash` and then execute the script as `./compress-folders.xsh` does the error still show up with the same frequency?
2. Can you try running on Python 3.8?  I doubt it's related but it will help disambiguate potential sources of error
3. If you start `xonsh` with `XONSH_DEBUG=1 xonsh` then you'll get a more informative traceback (it will disable the amalgamation that mangles the traceback)

**Comment by tex** [comment]
Mar 13 15:45:48 nixos-server systemd[1]: muj-android-photo-backup.service: Consumed 2.805s CPU time, received 833.9K IP traffic, sent 28.6K IP traffic.
I am running xsh script from systemd service. This error happens very frequently. Almost every other call.

```
Mar 13 15:45:48 nixos-server systemd[1]: Starting Milan's phone photo backup...
Mar 13 15:45:48 nixos-server xonsh[4320]: Traceback (most recent call last):
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/bin/.xonsh-wrapped", line 9, in <module>
Mar 13 15:45:48 nixos-server xonsh[4320]:     sys.exit(main())
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 24539, in main
Mar 13 15:45:48 nixos-server xonsh[4320]:     _failback_to_other_shells(args, err)
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 24486, in _failback_to_other_shells
Mar 13 15:45:48 nixos-server xonsh[4320]:     raise err
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 24537, in main
Mar 13 15:45:48 nixos-server xonsh[4320]:     sys.exit(main_xonsh(args))
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 24584, in main_xonsh
Mar 13 15:45:48 nixos-server xonsh[4320]:     run_script_with_cache(
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 3006, in run_script_with_cache
Mar 13 15:45:48 nixos-server xonsh[4320]:     run_compiled_code(ccode, glb, loc, mode)
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 2911, in run_compiled_code
Mar 13 15:45:48 nixos-server xonsh[4320]:     func(code, glb, loc)
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/home/milan/Pictures/MobilePhoneAndroid/get_pictures.xsh", line 3, in <module>
Mar 13 15:45:48 nixos-server xonsh[4320]:     cd /home/milan/Pictures/MobilePhoneAndroid/
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 22928, in subproc_captured_hiddenobject
Mar 13 15:45:48 nixos-server xonsh[4320]:     return run_subproc(cmds, captured="hiddenobject")
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 22887, in run_subproc
Mar 13 15:45:48 nixos-server xonsh[4320]:     command.end()
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 15703, in end
Mar 13 15:45:48 nixos-server xonsh[4320]:     self._end(tee_output=tee_output)
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 15711, in _end
Mar 13 15:45:48 nixos-server xonsh[4320]:     for _ in self.tee_stdout():
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 15624, in tee_stdout
Mar 13 15:45:48 nixos-server xonsh[4320]:     for line in self.iterraw():
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 15500, in iterraw
Mar 13 15:45:48 nixos-server xonsh[4320]:     stdout = NonBlockingFDReader(stdout.fileno(), timeout=timeout)
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 13792, in __init__
Mar 13 15:45:48 nixos-server xonsh[4320]:     super().__init__(fd, timeout=timeout)
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 13665, in __init__
Mar 13 15:45:48 nixos-server xonsh[4320]:     self.queue = queue.Queue()
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/mzhsnhly92hg1qd00nh4130rii33lz6i-xonsh-0.9.21/lib/python3.8/site-packages/xonsh/__amalgam__.py", line 78, in __getattribute__
Mar 13 15:45:48 nixos-server xonsh[4320]:     m = _import_module(mod)
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "/nix/store/xibghivp12jgk2xrwykpxxhy8wbmr5zi-python3-3.8.8/lib/python3.8/importlib/__init__.py", line 127, in import_module
Mar 13 15:45:48 nixos-server xonsh[4320]:     return _bootstrap._gcd_import(name[level:], package, level)
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "<frozen importlib._bootstrap_external>", line 779, in exec_module
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "<frozen importlib._bootstrap_external>", line 891, in get_code
Mar 13 15:45:48 nixos-server xonsh[4320]:   File "<frozen importlib._bootstrap_external>", line 972, in get_data
Mar 13 15:45:48 nixos-server xonsh[4320]: OSError: [Errno 9] Bad file descriptor
Mar 13 15:45:48 nixos-server systemd[1]: muj-android-photo-backup.service: Control process exited, code=exited, status=1/FAILURE
```

**Comment by anki-code** [comment]
Hi @tex! Do you have repeatable example? It will be great of it was in docker container.

**Comment by tex** [comment]
Hi @anki-code Well, yes, but I am not docker user. BTW I am having it on NixOS. I enabled $XONSH_DEBUG and not getting this error (I would say that I get it less frequently than before):

```
Mar 14 12:57:24 nixos-server xonsh[6264]: xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Mar 14 12:57:24 nixos-server xonsh[6264]: Traceback (most recent call last):
Mar 14 12:57:24 nixos-server xonsh[6264]:   File "/nix/store/7npdc0q78xxbqvgbfaz5lbkkywvs5fjq-xonsh-0.9.21/lib/python3.8/site-packages/>
Mar 14 12:57:24 nixos-server xonsh[6264]:     for f in os.listdir(data_dir)
Mar 14 12:57:24 nixos-server xonsh[6264]: OSError: [Errno 9] Bad file descriptor: '/home/milan/.local/share/xonsh'
Mar 14 12:57:24 nixos-server xonsh[6264]: Could not collect xonsh history files.
```

The folder `/home/milan/.local/share/xonsh` exists of-course, and script has access right there.

It's `python3-3.8.8` if it matters.

**Comment by anki-code** [comment]
@tex please do in xonsh:
```python
cd $XONSH_DATA_DIR
files = [os.path.join($XONSH_DATA_DIR, f) for f in os.listdir($XONSH_DATA_DIR) if f.startswith("xonsh-") and f.endswith(".json")]
len(files)
```
If you have success tell me the output and please give me more details what you're doing (how you run xonsh and for what). If you run xonsh in the loop probably this code is not stable for multiple runs and we should understand this.

If you'll have an error please do `ls $XONSH_DATA_DIR | wc -l` and tell me the output.

**Comment by tex** [comment]
Number #1: 1266
Number #2: 1266

My NixOS systemd service:
```
  systemd.services."muj-android-photo-backup" =
  { description = "Milan's phone photo backup";
          
    wantedBy = [ "multi-user.target" ];
          
    path = with pkgs; [ ffmpeg imagemagick jpegoptim ];
    
    environment = {
      "XONSH_DEBUG" = "1";
    };
 
    serviceConfig =
    { # StandardError = "journal";
      WorkingDirectory = "/home/milan/Pictures/MobilePhoneAndroid";
      ExecStart = ''  
        ${pkgs.xonsh.overrideAttrs(oldAttrs: {
          # I want to use pysftp in xonsh to easily work with sftp
          propagatedBuildInputs = oldAttrs.propagatedBuildInputs ++ [ pkgs.python3Packages.pysftp ];
        })}/bin/xonsh /home/milan/Pictures/MobilePhoneAndroid/get_pictures.xsh
      '';
      Type = "forking";
      User = "milan";
      Restart = "always";
      RestartSec = "5m";
    };
  };
```

My get_pictures.xsh script:

```
#!env xonsh --shell-type readline

$XONSH_SHOW_TRACEBACK = True

cd /home/milan/Pictures/MobilePhoneAndroid/

import pysftp
import os
import time

def create_image_thumbnail(rfilename, lfilename):
    print("Creating image thumbnail for {} at b/{}".format(rfilename, lfilename))
    # Create thumbnail and optimize it, run it on background
    convert @(rfilename) -resize 20% b/@(lfilename) && jpegoptim -s -S20% b/@(lfilename) &


def create_video_thumbnail(rfilename, lfilename):
    print("Creating video thumbnail for {} at b/{}".format(rfilename, lfilename))
    # Create smaller video thumbnail, run it on background
    ffmpeg -i @(rfilename) -vf scale=320:-2 b/@(lfilename) &


# Remove IMG_ and keep rest (date_time_num.jpg)...
# Remove VID_ and keep rest (date_time_num.mp4)...
def r2lfilename(rfilename):
    lfilename = rfilename
    if (rfilename.startswith('IMG_') or
            rfilename.startswith('VID_')):
        lfilename = rfilename[4:]
    return lfilename


def sync_one_file(rfilename, rsize, sftp):
    # Skip syncing if the file is already present
    # locally and has the correct file size.
    if (os.path.isfile(rfilename) and
            rsize == os.path.getsize(rfilename)):
        print("\tFile is already synced, skipping...")
        return

    # File is on remote, but not on locale
    # or file sizes differ => download file from
    # remote to locale (first time or again)
    sftp.get(rfilename, rfilename, preserve_mtime=True)

    # Confirm that file was successfully transferred.
    # If not, remove incorrect local copy and continue
    # with next file assuming that we get current file
    # next time this script is run.
    if (os.path.isfile(rfilename)
            and rsize != os.path.getsize(rfilename)):
        print(('\tFile was transferred but corrupted, '
               'removing wrong local copy...'))
        os.remove(rfilename)
        return

    # Create local file name used to store thumbnail.
    lfilename = r2lfilename(rfilename)

    if (rfilename.lower().endswith('.jpg')):
        create_image_thumbnail(rfilename, lfilename)
    elif (rfilename.lower().endswith('.mp4')):
        create_video_thumbnail(rfilename, lfilename)

print("Fetching and processing pictures and videos from phone has begun")

# No idea why it doesn't work with cnopts with reseted hostkeys,
# no idea why username in Connection is ignored and thus explicit
# private_key is needed.
cnopts = pysftp.CnOpts()
cnopts.hostkeys = None
sftp = pysftp.Connection('192.168.1.142', port=2222,
                         private_key='/home/milan/.ssh/id_rsa_android',
                         cnopts=cnopts)
sftp.chdir('/mnt/sdcard/DCIM/Camera')
rfiles = sftp.listdir_attr()
for rfile in rfiles:
    # Remote file name
    rfilename = rfile.filename
    # Remote file size
    rsize = rfile.st_size
    # Apply it only for .jpg and .mp4 files
    if (rfilename.lower().endswith('.jpg') or
            rfilename.lower().endswith('.mp4')):
        print("Syncing remote file: {}, {} B".format(rfilename, rsize))
        sync_one_file(rfilename, rsize, sftp)

# Wait until all thumbnails are done
while $(jobs):
    time.sleep(1)


import fnmatch
import os
import re

def findfiles(which, where='.'):
    '''Returns list of filenames from `where` path matched by 'which'
       shell pattern. Matching is case-insensitive.'''
    # TODO: recursive param with walk() filtering
    rule = re.compile(fnmatch.translate(which), re.IGNORECASE)
    return [name for name in os.listdir(where) if rule.match(name)]


# Create thumbnail for all pictures here that does not have it already
for rfilename in findfiles('*.jpg'):
    lfilename = r2lfilename(rfilename)
    if not pf'b/{lfilename}'.exists():
        create_image_thumbnail(rfilename, lfilename)
# Create thumbnail for all videos here that does not have it already
for rfilename in findfiles('*.mp4'):
    lfilename = r2lfilename(rfilename)
    if not pf'b/{lfilename}'.exists():
        create_video_thumbnail(rfilename, lfilename)

# Delete every picture at DCIM/Camera that we have here
sftp.chdir('/mnt/sdcard/DCIM/Camera')
rfiles = [rfile.filename for rfile in sftp.listdir_attr()]
lfiles = findfiles('*.jpg') + findfiles('*.mp4')
for lfile in lfiles:
    if lfile in rfiles:
        print('Removing {}'.format(lfile))
        sftp.remove(lfile)

# Wait until all thumbnails are made
while $(jobs):
    time.sleep(1)

# Copy thumbnails to DCIM/Camera/b
sftp.chdir('/mnt/sdcard/DCIM/Camera/b')
rfiles = [rfile.filename for rfile in sftp.listdir_attr()]
lfiles = findfiles('*.jpg', 'b') + findfiles('*.mp4', 'b')
for lfile in lfiles:
    if not lfile in rfiles:
        print('Copying thumbnail {}'.format(lfile))
        sftp.put('b/{}'.format(lfile))
```


Oh yes, xonsh is beautiful and I can't understand how I had been able to miss it for so long time :)

**Comment by anki-code** [comment]
@tex source code is good, thank you! But I mean the workflow of getting the error.
Example:
1. I run xonsh
2. Then I run xonsh script
3. I open new window, run xonsh and get the error

The error you've got is really strange because the code just get the list of files. This error means that the descriptor was closed before the end of reading.

**Comment by anki-code** [comment]
@tex If you can create a zip archive with code that repeat the error it will be great. But I understand that you have complex environment...

**Comment by tex** [comment]
@anki-code Well, my environment is pretty simple. You just need that script, pysftp and run it. It will raise ssh exception but that is OK, the problem with file descriptor is visible as it happens first when it happens. Seems possible that this error with file descriptor happens and then script runs anyway in the same "session", but I didn't look at that too deeply.
I run it from systemd service, with auto restart after 5minutes, but you can do shorter like 1minute.
I do not know 100% but I didn't see the error when I executed the script from command line when I was creating / debugging it.

Also it looks like that after adding `XONSH_HISTORY_BACKEND=dummy` into run environment solved the problem. Will monitor it over couple days and post result here.

**Comment by anki-code** [comment]
@tex I think you catch very fun thing! The xonsh reads the history not only when it runs in interactive mode but also in case of running the script.

When you do `XONSH_HISTORY_BACKEND=dummy` you just disable the history reading on starting up.

I think we should completely disable history in case of not interactive run the script - 4177

**Comment by tex** [comment]
So since I started using the `XONSH_HISTORY_BACKEND=dummy` I have not seen the file descriptor error.

Yeah, I am always the guy who hits very fun things :)

**Comment by anki-code** [comment]
@tex The fix is already in master :)

**Comment by cottrell** [comment]
Is this unrelated to https://github.com/xonsh/xonsh/issues/4224 ?

**Comment by anki-code** [comment]
I prefer to close this issue because this thread became the group of error reports from different parts of xonsh but with one error text ("bad file descriptor") and without any minimal repeatable examples.

* If you see the error when try to running xonsh scripts. This issue was fixed in #4178 and will be released after xonsh 0.9.27. Please check the code from master or from newer version.

* If you use aliases in the loops/pipes see #4224

* In case of doubts just create a new issue with minimal repeatable example of your issue
